Npm Dependency Installation Fails with `ERESOLVE unable to resolve dependency tree`

After we increase the version of `prettier` in our projects, developers sometimes cannot use spotless at all. They get lint errors on formatted files, informing them that the `npm install` failed:

```
There were 39 lint error(s), they must be fixed or suppressed.
  build/<path to SQL file>:LINE_UNDEFINED prettier-format(com.diffplug.spotless.npm.NpmProcessException) Running npm command 'NpmInstall in <project path>/build/spotless-prettier-node-modules-edd532637edcb6904b3d70deaa4ddaa0 [<project path>/.gradle/nodejs/node-v22.14.0-darwin-arm64/bin/npm install --no-audit --no-fund --prefer-offline]' failed with exit code: 1 
```

Reproducing the command spotless runs reveals the error by npm spotless is facing:

<details>
<summary><code>npm install</code> error</summary>


```
npm error code ERESOLVE
npm error ERESOLVE unable to resolve dependency tree
npm error
npm error While resolving: spotless-prettier@2.0.0
npm error Found: prettier@undefined
npm error node_modules/prettier
npm error   dev prettier@"3.4.2" from the root project
npm error
npm error Could not resolve dependency:
npm error peer prettier@"^3.0.3" from prettier-plugin-sql@0.18.1
npm error node_modules/prettier-plugin-sql
npm error   dev prettier-plugin-sql@"0.18.1" from the root project
npm error
npm error Fix the upstream dependency conflict, or retry
npm error this command with --force or --legacy-peer-deps
npm error to accept an incorrect (and potentially broken) dependency resolution.
npm error
npm error
npm error For a full report see:
npm error <user home>/.npm/_logs/2025-03-14T09_43_53_020Z-eresolve-report.txt
npm error A complete log of this run can be found in: <user home>/.npm/_logs/2025-03-14T09_43_53_020Z-debug-0.log
```

</details>

The root problem is that `npm` needs to update its package information, but doesn’t do that when `--prefer-offline` is set. Spotless [already has logic for such a case](https://github.com/diffplug/spotless/blob/8187042d19a1a7593f42f28160be8515c3016227/lib/src/main/java/com/diffplug/spotless/npm/NodeApp.java#L98), but [the detection for when to apply that logic](https://github.com/diffplug/spotless/blob/8187042d19a1a7593f42f28160be8515c3016227/lib/src/main/java/com/diffplug/spotless/npm/NodeApp.java#L113) does not include this case, yet.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -12,6 +12,7 @@ We adhere to the [keepachangelog](https://keepachangelog.com/en/1.0.0/) format (
 ## [Unreleased]
 ### Changed
 * Use palantir-java-format 2.57.0 on Java 21. ([#2447](https://github.com/diffplug/spotless/pull/2447))
+* Re-try `npm install` with `--prefer-online` after `ERESOLVE` error. ([#2448](https://github.com/diffplug/spotless/pull/2448))
 
 ## [3.1.0] - 2025-02-20
 ### Added

--- a/lib/src/main/java/com/diffplug/spotless/npm/NodeApp.java
+++ b/lib/src/main/java/com/diffplug/spotless/npm/NodeApp.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2023 DiffPlug
+ * Copyright 2023-2025 DiffPlug
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -98,18 +98,28 @@ public class NodeApp {
 			if (!offlineInstallFailed(e.getResult())) {
 				throw e; // pass through
 			}
-			// if the npm install fails with message "No matching version found for <package>@<version>", we try again without the offline flag
+			// if the npm install fails in a way that might be caused by missing dependency information, we try again without the offline flag
 			npmProcessFactory.createNpmInstallProcess(nodeServerLayout, formatterStepLocations, PREFER_ONLINE).waitFor();
 		}
 	}
 
-	private boolean offlineInstallFailed(ProcessRunner.Result result) {
+	private static boolean offlineInstallFailed(ProcessRunner.Result result) {
 		if (result == null) {
 			return false; // no result, something else must have happened
 		}
 		if (result.exitCode() == 0) {
 			return false; // all is well
 		}
-		return result.stdOutUtf8().contains("code ETARGET") && result.stdOutUtf8().contains("No matching version found for"); // offline install failed, needs online install
+		var installOutput = result.stdOutUtf8();
+		// offline install failed, needs online install
+		return isNoMatchingVersionFound(installOutput) || isCannotResolveDependencyTree(installOutput);
+	}
+
+	private static boolean isNoMatchingVersionFound(String installOutput) {
+		return installOutput.contains("code ETARGET") && installOutput.contains("No matching version found for");
+	}
+
+	private static boolean isCannotResolveDependencyTree(String installOutput) {
+		return installOutput.contains("code ERESOLVE") && installOutput.contains("unable to resolve dependency tree");
 	}
 }

--- a/plugin-gradle/CHANGES.md
+++ b/plugin-gradle/CHANGES.md
@@ -5,6 +5,7 @@ We adhere to the [keepachangelog](https://keepachangelog.com/en/1.0.0/) format (
 ## [Unreleased]
 ### Changed
 * Use palantir-java-format 2.57.0 on Java 21. ([#2447](https://github.com/diffplug/spotless/pull/2447))
+* Re-try `npm install` with `--prefer-online` after `ERESOLVE` error. ([#2448](https://github.com/diffplug/spotless/pull/2448))
 
 ## [7.0.2] - 2025-01-14
 ### Fixed

--- a/plugin-maven/CHANGES.md
+++ b/plugin-maven/CHANGES.md
@@ -5,6 +5,7 @@ We adhere to the [keepachangelog](https://keepachangelog.com/en/1.0.0/) format (
 ## [Unreleased]
 ### Changed
 * Use palantir-java-format 2.57.0 on Java 21. ([#2447](https://github.com/diffplug/spotless/pull/2447))
+* Re-try `npm install` with `--prefer-online` after `ERESOLVE` error. ([#2448](https://github.com/diffplug/spotless/pull/2448))
 
 ## [2.44.3] - 2025-02-20
 * Support for `clang-format` ([#2406](https://github.com/diffplug/spotless/pull/2406))

